### ISSUE TITLE
tlsdialer panics

### ISSUE THREAD

**Issue by fffw** [getlantern/lantern#3184]
When running on Android

```
E/Go      ( 7224): panic: runtime error: invalid memory address or nil pointer dereference
E/Go      ( 7224): [signal 0xb code=0x1 addr=0x18 pc=0x79c20604]
E/Go      ( 7224): goroutine 3365 [running]:
E/Go      ( 7224): crypto/tls.clientSessionCacheKey(0x0, 0x0, 0x8a70c780, 0x0, 0x0)
E/Go      ( 7224): 	/go/src/crypto/tls/handshake_client.go:643 +0x50
E/Go      ( 7224): crypto/tls.(*Conn).clientHandshake(0x8a70f380, 0x0, 0x0)
E/Go      ( 7224): 	/go/src/crypto/tls/handshake_client.go:107 +0xb0c
E/Go      ( 7224): crypto/tls.(*Conn).Handshake(0x8a70f380, 0x0, 0x0)
E/Go      ( 7224): 	/go/src/crypto/tls/conn.go:974 +0x110
E/Go      ( 7224): github.com/getlantern/tlsdialer.DialForTimings.func3(0x8adaa840, 0x8a70f380)
E/Go      ( 7224): 	/src/github.com/getlantern/tlsdialer/tlsdialer.go:180 +0x1c
E/Go      ( 7224): created by github.com/getlantern/tlsdialer.DialForTimings
E/Go      ( 7224): 	/src/github.com/getlantern/tlsdialer/tlsdialer.go:181 +0x14fc
```

**Comment by fffw** [comment]
Ok, that's beause `conn.RemoteAddr()` returns nil, and the evil code is detour.go#L339.

**Comment by fffw** [comment]
Weird, same panic still happens after Lantern Android launched for 3~4 min.

**Comment by atavism** [comment]
@fffw Hmm, could you forward me your full log? Does it happen consistently and always around the 3-4 minute mark?

**Comment by fffw** [comment]
I only run it twice so can't say it's consistent or not, but the only reason I can think of is that `c.conn.RemoteAddr()` at https://golang.org/src/crypto/tls/handshake_client.go#L107 returns `nil`.

Oh I just found the link in your email and PR are the same,so do the APKs I just downloaded, both 4184936 bytes. Any possibility it is cached? Maybe upload with a new name or mail directly @atavism ?

**Comment by atavism** [comment]
Ha, @fffw.. I figured out the cause of that "no X-LANTERN-AUTH-TOKEN field" error I was suddenly seeing in the ATS server log after merging the latest changes from valencia. I forgot we're using that hack, substituting "connect" with "tcp" as the transport argument, whenever we send the CONNECT request to the chained server. Should be fixed now. I just uploaded another APK.

**Comment by atavism** [comment]
Oh, and you were right. Looks like the shared library with your changes wasn't even integrated with the latest build. I started seeing that same tlsdialer panic. Want to try out the new version whenever you get a chance?

**Comment by atavism** [comment]
I'm still seeing this actually. Will keep investigating.

**Comment by atavism** [comment]
I confirmed ``RemoteAddr()`` is definitely returning nil whenever we attempt the handshake in ``tlsdialer`` (we're panicking because we attempt to dereference it). I've surmised the issue is related to how we're creating the protected connections that bypass the VpnService: these connections are created/destroyed rapidly and, under certain scenarios, an underlying socket could be closed immediately prior to dialing a connection. I resolved this by adding a check for a nil ``RemoteAddr()`` and returning an error before we attempt the ``Handshake``; the relevant changes are here: https://github.com/getlantern/lantern/commit/d1b171eee1557d59d7d3cf8a794a44d41dbd3dc2 - even when we do encounter this, in practice, we typically end up trying to reconnect almost immediately.
